packages/adblocker-webextension/src/index.ts

```diff
@@ -128,19 +140,110 @@ export function updateResponseHeadersWithCSP(
   return { responseHeaders };
 }
 
+const HTML_FILTERABLE_REQUEST_TYPES = new Set<RequestType>([
+  'main_frame',
+  'mainFrame',
+  'sub_frame',
+  'subFrame',
+  'stylesheet',
+  'script',
+  'document',
+  'fetch',
+  'prefetch',
+  'preflight',
+  'xhr',
+  'xmlhttprequest',
+]);
+
+// html filters are applied to text/* and those additional mime types
+const HTML_FILTERABLE_NON_TEXT_MIME_TYPES = new Set([
+  'application/javascript',
+  'application/json',
+  'application/mpegurl',
+  'application/vnd.api+json',
+  'application/vnd.apple.mpegurl',
+  'application/vnd.apple.mpegurl.audio',
+  'application/x-javascript',
+  'application/x-mpegurl',
+  'application/xhtml+xml',
+  'application/xml',
+  'audio/mpegurl',
+  'audio/x-mpegurl',
+]);
+export const MAXIMUM_RESPONSE_BUFFER_SIZE = 10 * 1024 * 1024;
+
+function getHeaderFromDetails(
+  details: OnHeadersReceivedDetailsType,
+  headerName: WebRequest.HttpHeadersItemType['name'],
+): string | undefined {
+  return details.responseHeaders?.find((header) => header.name === headerName)?.value;
+}
+
+// $replace filters are applied to complete response bodies
+// To avoid performance problem we should ignore large request or binary data
+export function shouldApplyReplaceSelectors(
+  request: Request<OnBeforeRequestDetailsType | OnHeadersReceivedDetailsType>,
+): boolean {
+  const details = request._originalRequestDetails as OnHeadersReceivedDetailsType;
+
+  // In case of undefined error of xhr/fetch and any kind of network activities
+  if (details.statusCode === 0) {
+    return false;
+  }
+
+  if (details.statusCode < 200 || details.statusCode > 299) {
```

Comment:
Shouldn't we handle 5xx and 4xx requests? 3xx are redirects but 5xx and 4xx still can include ad scripts.

Reply:
I don't see why 4xx or 5xx would include ads - those are error codes. Do we have any proves that ads are shipped that way?